packages/@aws-cdk/aws-stepfunctions/lib/state-machine.ts

```diff
@@ -345,7 +345,7 @@ abstract class StateMachineBase extends Resource implements IStateMachine {
 /**
  * Define a StepFunctions State Machine
  */
-export class StateMachine extends StateMachineBase {
+export class StateMachine extends StateMachineBase implements iam.IGrantable {
```

Comment:
this should be done on an `IStateMachine` and not here so that it will also work with imported state machines

Reply:
Is this a blocker for solving this issue? Or can I open another issue to work on a PR that moves IGrantable down to IStateMachine?

Implementing it on the StateMachine was straight forward, but to add it to the Interface I will need to investigate what other attributes should also be provided on import.

Reply:
For the imported one, you simply provide the `grantPrincipal` as `new iam.UnknownPrincipal({ resource: this });`. Here's an example from the existing libraries: https://github.com/aws/aws-cdk/blob/06b6d820b4ad7f913b8538bab63ce3a42af4be8f/packages/%40aws-cdk/aws-codebuild/lib/project.ts#L641-L641

Reply:
This won't allow granting access on an imported StateMachine though, right? For that I would need to provide a .fromAttributes() that takes an IRole among other things.

Reply:
what makes adding it to the `IStateMachine` difficult? it's implemented by `StateMachineBase` and I think we should be doing it there. 

Take a look at our [example construct library](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/example-construct-library/lib/example-resource.ts#L19-L141) for inspiration.

wdyt?

Reply:
> This won't allow granting access on an imported StateMachine though, right? For that I would need to provide a .fromAttributes() that takes an IRole among other things.

`grant*()` methods and the `IGrantable` interface are different things (although named similarly, so I definitely understand the confusion 😉).

`IGrantable` is used when the StateMachine's Role is being modified - that is, when the StateMachine itself is passed as the first argument to `grant*()` (something like `myBucket.grantRead(myStateMachine);`.

`grant*()` methods on the StateMachine itself have nothing to do with the `IGrantable` implementation inside the `StateMachine` class 🙂.

Reply:
Is this a blocker for solving this issue? Or can I open another issue to work on a PR that moves IGrantable down to IStateMachine?

Reply:
@skinny85 I'm perfectly aware of the difference. But `myBucket.grantRead(myStateMachine)` won't work on an imported `IStateMachine`, because the role is not defined. So I'm not quite seeing the point of implementing the IGrantable on `IStateMachine`. For that to make sense, we would also need to implement `.fromStateMachineAttributes()`. And for that, a whole bunch of other attributes should be supported. To implement that, I suggested to create a new issue.

Of course I can simply move the `IGrantable` to `IStateMachine` if that's a blocker (even if it doesn't actually provide any functionality). @shivlaks could you please comment on that?

Reply:
> @skinny85 I'm perfectly aware of the difference. But myBucket.grantRead(myStateMachine) won't work on an imported IStateMachine, because the role is not define

That's fine - that's exactly the situation today. So moving `IGrantable` to `IStateMachine` does not substantially change that.

I would do that, and then perhaps follow-up with a different PR that adds `fromStateMachineAttributes()`.